File: github/branch.go
package github

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/github/hub/git"
)

type Branch struct {
	Repo *GitHubRepo
	Name string
}

var (
	shortRemoteRe = regexp.MustCompile("^refs/(remotes/)?.+?/")
	longRemoteRe  = regexp.MustCompile("^refs/(remotes/)?")
	remoteRe      = regexp.MustCompile("^refs/remotes/([^/]+)")
)

func (b *Branch) ShortName() string {
	return shortRemoteRe.ReplaceAllString(b.Name, "")
}

func (b *Branch) LongName() string {
	return longRemoteRe.ReplaceAllString(b.Name, "")
}

func (b *Branch) RemoteName() string {
	if remoteRe.MatchString(b.Name) {
		return remoteRe.FindStringSubmatch(b.Name)[1]
	}

	return ""
}

func (b *Branch) Upstream() (u *Branch, err error) {
	name, err := git.SymbolicFullName(fmt.Sprintf("%s@{upstream}", b.ShortName()))
	if err != nil {
		return
	}

	u = &Branch{b.Repo, name}

	return
}

func (b *Branch) IsMaster() bool {
	masterName := b.Repo.MasterBranch().ShortName()
	return b.ShortName() == masterName
}

func (b *Branch) IsRemote() bool {
	return strings.HasPrefix(b.Name, "refs/remotes")
}

File: lib/hub/context/branch.rb
module Hub
  module Context
    class Branch < Struct.new(:local_repo, :name)
      alias to_s name

      def short_name
        name.sub(%r{^refs/(remotes/)?.+?/}, '')
      end

      def master?
        master_name = if local_repo then local_repo.master_branch.short_name
                      else 'master'
                      end
        short_name == master_name
      end

      def upstream
        if branch = local_repo.git_command("rev-parse --symbolic-full-name #{short_name}@{upstream}")
          Branch.new local_repo, branch
        end
      end

      def remote?
        name.index('refs/remotes/') == 0
      end

      def remote_name
        name =~ %r{^refs/remotes/([^/]+)} and $1 or
          raise Error, "can't get remote name from #{name.inspect}"
      end
    end
  end
end
